### PR TITLE
JENKINS-52396: Support for Perforce in Blueocean

### PR DESCRIPTION
**Pull request by p4charu**
# Description
Support for Perforce in Blueocean. This is a functionality for creating a multi-branch pipeline for a Swarm project using Blueocean.
Needs review from the blueocean community. @blueoceandeploy
See https://issues.jenkins-ci.org/browse/JENKINS-52396
# Submitter checklist
I'm working on the javascript unit tests. Nonetheless, it's manually tested and works fine.

### CODE REVIEW

**Comment by p4charu** [comment]
Ready for review. @blueoceandeploy 
Support for Perforce in Blueocean
[https://issues.jenkins-ci.org/browse/JENKINS-52396](url)

File: pom.xml [hunk 1]
```diff
@@ -478,6 +478,13 @@
             <version>4.5.5-3.0</version>
         </dependency>
 
+        <!-- p4 scm plugin -->
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>p4</artifactId>
```
Comment (halkeye) [inline pom.xml]: I don't actually see any new java code, so is this actually a requirement?

Comment (p4charu) [reply pom.xml]: You are right. Took that one out.

**Comment by halkeye** [comment]
can you rebase? hopefully this'll mean it now passes the tests

**Comment by p4charu** [comment]
Updated my fork. All tests pass now. Thanks.

**Comment by halkeye** [comment]
Sweet. I haven't had a chance to pull the branch and try it out.

Are there free p4 servers I can run to test this? 

As much I've been fighting with the tests (mostly e2e) i still think this PR would benefit from having them. Do you want to add at least some basic tests for the frontend? Make sure it compiles/renders and stuff?

**Comment by p4paul** [comment]
Hi Gavin, you can use an unlicensed Perforce server for free, but it is limited to 5 users (I think that should be enough for testing).

You can download the p4 (client) and p4d (server) binaries from our FTP site:
   ftp://ftp.perforce.com/perforce/r19.1/

Or you could try my docker demos which use linux packages (specifically 'helix-p4d') .
   https://github.com/p4paul/helix-docker

I have not tried the docker demos for a while, so they may not work with the latest packages.

**Comment by p4charu** [comment]
> Sweet. I haven't had a chance to pull the branch and try it out.
> 
> Are there free p4 servers I can run to test this?
> 
> As much I've been fighting with the tests (mostly e2e) i still think this PR would benefit from having them. Do you want to add at least some basic tests for the frontend? Make sure it compiles/renders and stuff?

Added a basic test for PerforceCredentialsManager. Will add some more tests.

**Comment by p4charu** [comment]
Added tests for PerforceFlowManager

**Comment by halkeye** [comment]
image

What kind of credentials does it need? The dropdown is empty an doesn't tell me what to do next.
The other ones let you add a credential, so can you at least add a note to what type if its not a normal type?

**Comment by halkeye** [comment]
So I tried both your version and https://hub.docker.com/r/noonien/perforce-server/dockerfile

When I try to access /swarm/projects/?credential=admin i get "java.net.MalformedURLException: no protocol: null/api/version"

Which port should I be connecting to?

**Comment by benholtz** [comment]
Any update here?  There are many here waiting on the sidelines for this functionality.  Thank you!

**Comment by benholtz** [comment]
@p4charu @p4paul @halkeye @olamy @vivek @imeredith @sophistifunk @scherler @jennbriden @NicuPascu @kzantow 
Can you help get this merged in?  I know there are a great deal of perforce users that could use this functionality including myself.

**Comment by halkeye** [comment]
1) I'm not involved with this plugin anymore
2) tests are not passing
3) I asked about tearing with no response

If you want to get it merged the branch needs to be fixed up at a minimum

**Comment by ravivedula** [comment]
> Added tests for PerforceFlowManager
Hi, Not sure who the person responsible for this. If you are still responsible for this, can you update on when this is likely to be fixed and released? If you are not responsible, kindly guide me on whom to approach. Thank you and appreciate your early response.

**Comment by ravivedula** [comment]
@p4charu @halkeye @p4paul 
Hi, Not sure who the person responsible for this. If you are still responsible for this, can you update on when this is likely to be fixed and released? If you are not responsible, kindly guide me on whom to approach. Thank you and appreciate your early response.

**Comment by halkeye** [comment]
It'll be likely released once it's fixed. It'll be fixed when someone has time and interest to do so.

**Comment by halkeye** [comment]
I'm locking and closing this. If it gets worked on again it can get reopened.

**Pull request closed without merging**